pyiceberg/table/metadata.py

```diff
@@ -313,6 +315,34 @@ def construct_partition_specs(cls, data: Dict[str, Any]) -> Dict[str, Any]:
 
         return data
 
+    @model_validator(mode="before")
+    def construct_v1_spec_from_v2_fields(cls, data: Dict[str, Any]) -> Dict[str, Any]:
```

Comment:
I'd rather avoid the before validators to patch in the partition-spec and schema (probably sort-order as well?). I'd much rather see that we explicitly pass in the spec below:

```python
    if format_version == 1:
        return TableMetadataV1(
            location=location,
            schema=fresh_schema,
            last_column_id=fresh_schema.highest_field_id,
            current_schema_id=fresh_schema.schema_id,
+           partition_spec=fresh_partition_spec,
            partition_specs=[fresh_partition_spec],
            default_spec_id=fresh_partition_spec.spec_id,
+           sort_order=fresh_sort_order,
            sort_orders=[fresh_sort_order],
            default_sort_order_id=fresh_sort_order.order_id,
            properties=properties,
            last_partition_id=fresh_partition_spec.last_assigned_field_id,
            table_uuid=table_uuid,
        )
```

My biggest concern is again that this is a before validator, and we don't take into account the aliases (`partition-specs`, `partition-spec-id`, etc).

It looks like Java just ignores the `schema` if there is a `schemas`: https://github.com/apache/iceberg/blob/5f577f1b9902ffe6181897a439686a31fc81b89a/core/src/main/java/org/apache/iceberg/TableMetadataParser.java#L343-L368

Reply:
Yeah it just felt a bit awkward to pass in the fields in a redundant way. As long as the source of truth is correct when reading the fields, which it is as you saw for `schemas` then we're good. I've updated.

Also PartitionSpec on the V1 metadata is a list[dict[str, any]], right now I work with that to preserve compatibility but that does make things a bit awkward since we have to pass in the model_dump of partition spec instead of just the normal `PartitionSpec`.

Reply:
Actually @HonahX had a good catch here, it's not just a dump of partition spec, the List[dict[str, any]] is a list of the fields themselves.